Flexible switching with rails 5.2

derosm2 opened the pull request:
Based off of [this fork with thread safe fixes](https://github.com/mikecmpbll/apartment/pull/1), so please review those changes as well!

brownbeagle committed:
Kernel#quietly was deprecated in Rails 4.2

meganemura committed:
Removed task dependency

kuzukuzu committed:
enhance db:drop task

IngusSkaistkalns committed:
Support ActiveRecord-5.2 migration context

meganemura committed:
Use guard clause

meganemura committed:
Swallow the AR::NoDatabaseError

Printing an Error message is a little bit annoying until invoke db:create.

# Conflicts:
#	lib/apartment/railtie.rb

derosm2 committed:
Call drop database for postgres

derosm2 committed:
No longer set default tenant to a hash

derosm2 committed:
Fix calls to connection_switch! method

derosm2 committed:
Always use postgresql adapter

derosm2 committed:
Rescue from expected errors when dropping a database

derosm2 committed:
Rescue expected errors when creating tenant

derosm2 committed:
Account for different ports and fallback to postgres db

Use postgres db as default so that we can create/drop tenant databases as needed

derosm2 committed:
Actually switch the connection

derosm2 committed:
Fix seed rake task

derosm2 committed:
Clear connections after rake tasks

Old connections are persisting among rake tasks called during db:reset

derosm2 committed:
Add public method for cleaning connections

- Useful for when we want to reset our state, but hold on to the config

derosm2 committed:
Always switch to thread specific connection when possible

- Fixes some issues I was seeing with tests
- Add a guard to return excluded models spec name

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -30,8 +30,11 @@ def create(tenant)
+              default_db = "postgres"
+              temp_config = config.dup
+              temp_config[:database] = default_db
+              connection_switch!(temp_config)
```
The old code was dropping the `schema_search_path`. Is that something that the new code should be doing as well?

derosm2 replied:
That doesn't apply to us since we never set (or change) the schema search path. That functionality is for having multiple tenants on the same database separated with a different schema.

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -30,8 +30,11 @@ def create(tenant)
-            if difference[:host]
-              connection_switch(config, without_keys: [:database, :schema_search_path])
+            if difference[:host] || difference[:port]
+              default_db = "postgres"
```
Why is this in a var?

derosm2 replied:
I was contemplating abstracting this out to a config. Probably makes sense to just remove the var.

derosm2 replied:
Removed in 1e9ca1f0de305db33f845fc20daf170add4ec61b

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -54,8 +59,11 @@ def drop(tenant)
-        if difference[:host]
-          connection_switch(config, without_keys: [:database])
+        if difference[:host] || difference[:port]
+          default_db = "postgres"
```
Why is this in a var?

derosm2 replied:
Removed in 1e9ca1f0de305db33f845fc20daf170add4ec61b

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -54,8 +59,11 @@ def drop(tenant)
+          default_db = "postgres"
+          temp_config = config.dup
+          temp_config[:database] = default_db
+          connection_switch!(temp_config)
```
Should this be `connection_switch` without the `!`?

derosm2 replied:
There is no `connection_switch`, only `connection_switch!`.

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -116,12 +126,10 @@ def setup_connection_specification_name
+            return :_apartment_excluded if @connection_specification_name == :_apartment_excluded
+
+            return Thread.current[:_apartment_connection_specification_name] ||
+              (self == ActiveRecord::Base ? "primary" : superclass.connection_specification_name)
```
Should this be `self.class == ActiveRecord::Base` ?

derosm2 replied:
No, this is overwriting an [existing method in rails](https://github.com/rails/rails/blob/5-2-stable/activerecord/lib/active_record/connection_handling.rb#L98).

Original implementation:
```rb
def connection_specification_name
	if !defined?(@connection_specification_name) || @connection_specification_name.nil?
		return self == Base ? "primary" : superclass.connection_specification_name
	end
	@connection_specification_name
end
```

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -116,12 +126,10 @@ def setup_connection_specification_name
-                (self == ActiveRecord::Base ? "primary" : superclass.connection_specification_name)
-            end
-            @connection_specification_name
+            return :_apartment_excluded if @connection_specification_name == :_apartment_excluded
```
The old code had a path to return `@connection_specification_name` if it existed where this new one doesn't. Was that just not necessary anymore?

derosm2 replied:
I ended up removing that path to fix issues with our test suite. The spec name was getting set (and cached in that instance variable) as `primary`, which I believe was then preventing connections to the database specified by the Thread's `_apartment_connection_specification_name` value.

derosm2 replied:
It could have been an issue with the the timing of how Apartment was being loaded, but I was unable to see anything obvious.

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -140,6 +148,7 @@ def connection_switch!(config, without_keys: [])
 
         unless Apartment.connection_handler.retrieve_connection_pool(config[:name])
           Apartment.connection_handler.establish_connection(config)
+          Apartment.establish_connection(config)
```
Either this architecture is unfortunate, or the naming here is off, but this feels redundant.

derosm2 replied:
So this hack was unfortunately the solution I came to when I realized that Apartment was not always switching databases when I expected it to. It's possible that some of the other changes I've since made to `connection_specification_name` would allow us to only call one `establish_connection` here.

AJMiller replied:
is it worth trying to resolve?

derosm2 replied:
Removed in 26ace7cba1a1c2c045fb78aaeaa56ca40911fada. Tested it out briefly and and things seem ok.

AJMiller commented on lib/apartment/adapters/postgresql_adapter.rb:
```diff
@@ -78,11 +75,16 @@ def connection_specification_name(config)
           "_apartment_#{config.hash}".to_sym
         else
-          host_hash = Digest::MD5.hexdigest(config[:host] || config[:url] || "127.0.0.1")
+          value = "#{config[:host]}:#{config[:port]}" || config[:url] || "127.0.0.1"
```
Should `url` override `host/port` here to follow the rails pattern?

AJMiller replied:
Also wondering if there should be a `config[:host]` option like there is in the original code. That, of course, would make this a little more complicated to calculate.

derosm2 replied:
We end up setting `pool_per_config` to true, so these changes aren't relevant anymore.

AJMiller replied:
should we revert these changes then or is there a reason to keep them?

derosm2 replied:
It could potentially help us locally if we ever decided to turn off `pool_per_config`, since we run different postgres servers on different ports.

AJMiller commented on lib/apartment/tenant.rb:
```diff
@@ -25,9 +25,7 @@ def init
-        config = Apartment.default_tenant
-
-        adapter_name = "#{config[:adapter]}_adapter"
+        adapter_name = "postgresql_adapter"
```
What was wrong with the old code here? It looks to be more flexible?

derosm2 replied:
The old code assumes that the default tenant is a hash, instead of a tenant name. We could pull up the config for the default tenant, but I went with this since we only ever need the postgresql adapter.

AJMiller commented on lib/apartment/adapters/abstract_adapter.rb:
```diff
@@ -164,9 +174,25 @@ def load_or_abort(file)
         raise TenantNotFound, "Error while connecting to tenant #{tenant}: #{exception.message}"
       end
+
+      def raise_create_tenant_error!(tenant, exception)
```
I'm a bit confused why the original fork dropped this code?

derosm2 replied:
🤷‍♂️ Some of the tenant create/drop actions weren't fully fleshed out in that fork. May have been that [this commit](https://github.com/mikecmpbll/apartment/pull/1/commits/6431e58c31207bfa23b11f38711afc84d7a0b8f7) just rewrote AbstractAdapter completely and didn't get around to bringing anything similar back in.

derosm2 committed:
Remove default_db var

derosm2 committed:
Add database config resolver

derosm2 committed:
Remove extra call to establish connection

derosm2 committed:
Call init again to ensure setup_connection_specification_name is called

- If the default tenant db does not exist, setup_connection_specification_name was not getting called
  - Other tenants would fail to get created as a result because the switching mechanism was not triggering

AJMiller reviewed (approved):
Although we agree it isn't ideal, this is a fair compromise that meets our needs right now. Ideally, I think we have a hardcoded connection to the `postgres` db that happens on init so we don't have to deal with the db not being found initially.

muyiwaoyeniyi commented:
@derosm2 Moving the conversation from https://github.com/influitive/apartment/issues/438 over here. 

I'm running rails 5.2.3 and ruby 2.6.1.

My config file isn't set up yet as I can't run generate apartment:install. Once I installed the gem from the branch, my app is unable to start or run any rails commands. I created the config file manually and tried to set it up but it doesn't appear to make a difference.

My question really is if there is a different setup or config that I need to get this gem working or the same config options from the original repo woks? I have a situation where I need to support tenants on different hosts running postgres db.

I tried config.tenant_resolver = Apartment::Resolvers::Database and use_schema options per the documentation and that's doesn't work. 

Thanks!

muyiwaoyeniyi commented:
@derosm2 Just checking if I could please get some help. Thanks!

derosm2 commented:
> @derosm2 Just checking if I could please get some help. Thanks!

@muyiwaoyeniyi What issues are you seeing when you try to run the rails commands? Also are you able to use Rails 6?

muyiwaoyeniyi commented:
@derosm2 Thanks for responding. I haven't tested with Rails 6. I'm running rails 5.2.3 and ruby 2.6.1. When I start my rails server, I get the error in the attachment below. It is coming from this file - abstract_adapter.rb:81 ...
![Screenshot from 2019-07-06 12-45-09](https://user-images.githubusercontent.com/1622600/60759031-93116600-9fec-11e9-98fc-b85a8e02f129.png)

When I try running - bundle exec rails generate apartment:install - I get basically the same error and the command hangs
![Screenshot from 2019-07-06 12-46-27](https://user-images.githubusercontent.com/1622600/60759035-a6bccc80-9fec-11e9-8606-d05460fb2324.png)

And so I think there is a configuration you might have on your end that I'm missing which is causing the error. Could be related to this https://github.com/derosm2/apartment/pull/1#pullrequestreview-216714813 ? 

This is what I have in my gemfile:
gem 'apartment', git: 'https://github.com/PatientWisdom/apartment.git', branch: 'flexible-switching-rails-5-2'

Thanks for the help!

derosm2 commented:
@muyiwaoyeniyi Looks like it's entering an infinite loop because it trying to connect to the default tenant, but there isn't one. Might be worth starting with an example config from the main repo
```rb
# You can have Apartment route to the appropriate Tenant by adding some Rack middleware.
# Apartment can support many different "Elevators" that can take care of this routing to your data.
# Require whichever Elevator you're using below or none if you have a custom one.
#
# require 'apartment/elevators/generic'
# require 'apartment/elevators/domain'
require 'apartment/elevators/subdomain'
# require 'apartment/resolvers/schema'

#
# Apartment Configuration
#
Apartment.configure do |config|
  # Add any models that you do not want to be multi-tenanted, but remain in the global (public) namespace.
  # A typical example would be a Customer or Tenant model that stores each Tenant's information.
  #
  # config.excluded_models = %w{ Tenant }

  # In order to migrate all of your Tenants you need to provide a list of Tenant names to Apartment.
  # You can make this dynamic by providing a Proc object to be called on migrations.
  # This object should yield either:
  # - an array of strings representing each Tenant name.
  # - a hash which keys are tenant names, and values custom db config (must contain all key/values required in database.yml)
  #
  # config.tenant_names = lambda{ Customer.pluck(:tenant_name) }
  # config.tenant_names = ['tenant1', 'tenant2']
  # config.tenant_names = [
  #   {
  #     adapter: 'postgresql',
  #     host: 'some_server',
  #     port: 5555,
  #     database: 'postgres' # this is not the name of the tenant's db
  #                          # but the name of the database to connect to before creating the tenant's db
  #                          # mandatory in postgresql
  #     schema_search_path: '"tenant1"'
  #   },
  #   'tenant2' => {
  #     adapter:  'postgresql',
  #     database: 'postgres' # this is not the name of the tenant's db
  #                          # but the name of the database to connect to before creating the tenant's db
  #                          # mandatory in postgresql
  #     
  #   }
  # }
  #
  config.tenant_names = lambda { ToDo_Tenant_Or_User_Model.pluck :database }

  # The tenant decorator setting should be a callable which receives the tenant
  # as an argument, and returns the a modified version of the tenant name which
  # you want to use in the resolver as a database or schema name, for example.
  #
  # A typical use-case might be prepending or appending the rails environment,
  # as shown below.
  #
  # config.tenant_decorator = ->(tenant){ "#{Rails.env}_#{tenant}" }

  # The resolver is used to convert a tenant name into a full spec. The two
  # provided resolvers are Database and Schema. When you issue
  # Apartment.switch("some_tenant"){ ... }, Apartment passes "some_tenant" to
  # the selected resolver (after it's been decorated). The Database resolver
  # takes the decorated tenant name, and inserts it into the :database key of a
  # full connection specification (the full spec is whatever the database spec
  # was at Apartment initialization. The schema resolver, does the same but
  # for the :schema_search_path option in the configuration.
  #
  # config.tenant_resolver = Apartment::Resolvers::Schema
end

# Setup a custom Tenant switching middleware. The Proc should return the name of the Tenant that
# you want to switch to.
# Rails.application.config.middleware.use 'Apartment::Elevators::Generic', lambda { |request|
#   request.host.split('.').first
# }

# Rails.application.config.middleware.use 'Apartment::Elevators::Domain'
Rails.application.config.middleware.use 'Apartment::Elevators::Subdomain'
# You can have Apartment route to the appropriate Tenant by adding some Rack middleware.
# Apartment can support many different "Elevators" that can take care of this routing to your data.
# Require whichever Elevator you're using below or none if you have a custom one.
#
# require 'apartment/elevators/generic'
# require 'apartment/elevators/domain'
require 'apartment/elevators/subdomain'
# require 'apartment/resolvers/schema'

#
# Apartment Configuration
#
Apartment.configure do |config|
  # Add any models that you do not want to be multi-tenanted, but remain in the global (public) namespace.
  # A typical example would be a Customer or Tenant model that stores each Tenant's information.
  #
  # config.excluded_models = %w{ Tenant }

  # In order to migrate all of your Tenants you need to provide a list of Tenant names to Apartment.
  # You can make this dynamic by providing a Proc object to be called on migrations.
  # This object should yield either:
  # - an array of strings representing each Tenant name.
  # - a hash which keys are tenant names, and values custom db config (must contain all key/values required in database.yml)
  #
  # config.tenant_names = lambda{ Customer.pluck(:tenant_name) }
  # config.tenant_names = ['tenant1', 'tenant2']
  # config.tenant_names = [
  #   {
  #     adapter: 'postgresql',
  #     host: 'some_server',
  #     port: 5555,
  #     database: 'postgres' # this is not the name of the tenant's db
  #                          # but the name of the database to connect to before creating the tenant's db
  #                          # mandatory in postgresql
  #     schema_search_path: '"tenant1"'
  #   },
  #   'tenant2' => {
  #     adapter:  'postgresql',
  #     database: 'postgres' # this is not the name of the tenant's db
  #                          # but the name of the database to connect to before creating the tenant's db
  #                          # mandatory in postgresql
  #     
  #   }
  # }
  #
  config.tenant_names = lambda { ToDo_Tenant_Or_User_Model.pluck :database }

  # The tenant decorator setting should be a callable which receives the tenant
  # as an argument, and returns the a modified version of the tenant name which
  # you want to use in the resolver as a database or schema name, for example.
  #
  # A typical use-case might be prepending or appending the rails environment,
  # as shown below.
  #
  # config.tenant_decorator = ->(tenant){ "#{Rails.env}_#{tenant}" }

  # The resolver is used to convert a tenant name into a full spec. The two
  # provided resolvers are Database and Schema. When you issue
  # Apartment.switch("some_tenant"){ ... }, Apartment passes "some_tenant" to
  # the selected resolver (after it's been decorated). The Database resolver
  # takes the decorated tenant name, and inserts it into the :database key of a
  # full connection specification (the full spec is whatever the database spec
  # was at Apartment initialization. The schema resolver, does the same but
  # for the :schema_search_path option in the configuration.
  #
  # config.tenant_resolver = Apartment::Resolvers::Schema
end

# Setup a custom Tenant switching middleware. The Proc should return the name of the Tenant that
# you want to switch to.
# Rails.application.config.middleware.use 'Apartment::Elevators::Generic', lambda { |request|
#   request.host.split('.').first
# }

# Rails.application.config.middleware.use 'Apartment::Elevators::Domain'
Rails.application.config.middleware.use 'Apartment::Elevators::Subdomain'
```

muyiwaoyeniyi commented:
@derosm2 I will look at it again. I believe I set it up correctly but maybe not. I'll let you know.

Thanks.

kaspernj commented:
I am running out of database connections. Is there a way of always establishing a single connect for each thread and not use a connection pool when connecting to tenant databases? I have a max connections in Postgres of 800 and I experience fork issues when raising it to 1000.

derosm2 commented:
> I am running out of database connections. Is there a way of always establishing a single connect for each thread and not use a connection pool when connecting to tenant databases? I have a max connections in Postgres of 800 and I experience fork issues when raising it to 1000.

This branch should obey the pool size and reuse connections where possible (and not create more pools). Not sure I'm following your situation exactly. can you provide some more details?
